File: python3/Intervals/largest_overlap_of_intervals.py
from ds import Interval
from typing import List

"""
Definition of Interval:
class Interval:
    def __init__(self, start: int, end: int):
        self.start = start
        self.end = end
"""


def largest_overlap_of_intervals(intervals: List[Interval]) -> int:
    points = []
    for interval in intervals:
        points.append((interval.start, 'S'))
        points.append((interval.end, 'E'))
    # Sort in chronological order. If multiple points occur at the same 
    # time, ensure end points are prioritized before start points.
    points.sort(key=lambda x: (x[0], x[1]))
    active_intervals = 0
    max_overlaps = 0
    for time, point_type in points:
        if point_type == 'S':
            active_intervals += 1
        else:
            active_intervals -= 1
        max_overlaps = max(max_overlaps, active_intervals)
    return max_overlaps

File: csharp/Intervals/LargestOverlapOfIntervals.cs
/*
	Definition of Interval:
	class Interval
    {
        public Interval(int start, int end)
        {
            Start = start; 
            End = end;
        }

        public int Start { get; set; }

        public int End { get; set; }
    }
*/

public class Solution
{
    public int LargestOverlapOfIntervals(Interval[] intervals)
    {
        List<(int time, char pointType)> points = [];

        foreach (Interval interval in intervals)
        {
            points.Add((interval.Start, 'S'));
            points.Add((interval.End, 'E'));
        }

        // Sort in chronological order. If multiple points occur at the same 
        // time, ensure end points are prioritized before start points.
        points.Sort((a, b) => a.time == b.time ? a.pointType.CompareTo(b.pointType) : a.time - b.time);

        int activeIntervals = 0, maxOverlaps = 0;

        foreach ((int time, int pointType) in points)
        {
            if (pointType == 'S')
                activeIntervals++;
            else
                activeIntervals--;

            maxOverlaps = Math.Max(maxOverlaps, activeIntervals);
        }

        return maxOverlaps;
    }
}
